handler/handler.go

```diff
@@ -2,47 +2,206 @@ package handler
 )
 
 var _ event.Handler = (*BatchHandler)(nil)
 
+var (
+	syncWaitGroup sync.WaitGroup
+
+	countChannel  = make(chan int)
+	insertChannel = make(chan int)
+	skipChannel   = make(chan int)
+	semaphore     = make(chan int, 5)
+)
+
 // BatchHandler handles batches of SearchDataImportModel events that contain CSV row data.
 type BatchHandler struct {
+	ElasticSearchCli esclient.Client
 }
 
 // NewBatchHandler returns a BatchHandler.
-func NewBatchHandler() *BatchHandler {
+func NewBatchHandler(ElasticSearchCli esclient.Client) *BatchHandler {
+
+	return &BatchHandler{
+		ElasticSearchCli: ElasticSearchCli,
+	}
+}
 
-	return &BatchHandler{}
+type esBulkResponse struct {
+	Took   int                  `json:"took"`
+	Errors bool                 `json:"errors"`
+	Items  []esBulkItemResponse `json:"items"`
+}
+
+type esBulkItemResponse map[string]esBulkItemResponseData
+
+type esBulkItemResponseData struct {
+	Index  string `json:"_index"`
+	Status int    `json:"status"`
+	Error  string `json:"error"`
 }
 
 // Handle the given slice of SearchDataImport Model.
 func (batchHandler BatchHandler) Handle(ctx context.Context, events []*models.SearchDataImportModel) error {
 	log.Info(ctx, "events handler called")
 
-	//TODO : temp output - integrating with Elastic search is addressed in separate task
-	for i := 0; i < len(events); i++ {
-		dataType := fmt.Sprintf("SearchDataImportModel.DataType, %s!", events[i].DataType)
-		searchIndex := fmt.Sprintf("SearchDataImportModel.SearchIndex, %s!", events[i].SearchIndex)
-		title := fmt.Sprintf("SearchDataImportModel.Title, %s!", events[i].Title)
-		keywords := fmt.Sprintf("SearchDataImportModel.Keywords, %s!", events[i].Keywords)
-		traceId := fmt.Sprintf("SearchDataImportModel.TraceId, %s!", events[i].TraceID)
-
-		logData := log.Data{
-			"dataType":    dataType,
-			"searchIndex": searchIndex,
-			"title":       title,
-			"keywords":    keywords,
-			"traceId":     traceId,
-		}
-		log.Info(ctx, "Event in Pipeline to be processed for Elastic Search", log.INFO, logData)
+	//no events receeived. Nothing more to do.
+	if len(events) == 0 {
+		log.Info(ctx, "there are no events to handle")
+		return nil
+	}
+	go status(ctx)
+
+	// This will block if we've reached our concurrency limit (sem buffer size)
+	err := batchHandler.SendToES(ctx, events, len(events))
+	if err != nil {
+		log.Fatal(ctx, "failed to send event to Elastic Search", err)
+		return err
 	}
 
+	time.Sleep(4 * time.Second)
+	syncWaitGroup.Wait()
+
 	log.Info(ctx, "event successfully handled")
 	return nil
 }
+
+func (bh BatchHandler) SendToES(ctx context.Context, events []*models.SearchDataImportModel, length int) error {
+
+	// Wait on semaphore if we've reached our concurrency limit
+	syncWaitGroup.Add(1)
+	semaphore <- 1
+
+	// Get Config
+	cfg, err := config.Get()
+	if err != nil {
+		log.Fatal(ctx, "error getting config", err)
+		os.Exit(1)
+	}
+
+	esDestURL := cfg.ElasticSearchAPIURL
+	esIndex := "esIndex"
```

Comment:
why are esIndex and esDestType hardcoded?

Reply:
* used esDestURL passed from the consumer config
* however, for esIndex (now changed as const "ons"), we need value for make a call to BulkUpdate (in dp-elasticseaerch lib)